staging/src/k8s.io/client-go/util/csaupgrade/upgrade_test.go

```diff
@@ -0,0 +1,98 @@
+package csaupgrade_test
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/util/csaupgrade"
+)
+
+func TestUpgradeCSA(t *testing.T) {
+	// Initial object has managed fields from using CSA
+	originalYAML := []byte(`
+apiVersion: v1
+data:
+  key: value
+  legacy: unused
+kind: ConfigMap
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","data":{"key":"value","legacy":"unused"},"kind":"ConfigMap","metadata":{"annotations":{},"name":"test","namespace":"default"}}
+  creationTimestamp: "2022-08-22T23:08:23Z"
+  managedFields:
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:data:
+        .: {}
+        f:key: {}
+        f:legacy: {}
+      f:metadata:
+        f:annotations:
+          .: {}
+          f:kubectl.kubernetes.io/last-applied-configuration: {}
+    manager: kubectl-client-side-apply
+    operation: Update
+    time: "2022-08-22T23:08:23Z"
+  name: test
+  namespace: default
+  resourceVersion: "12502"
+  uid: 1f186675-58e6-4d7b-8bc5-7ece581e3013
+`)
+	initialObject := unstructured.Unstructured{}
+	err := yaml.Unmarshal(originalYAML, &initialObject)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	upgraded, err := csaupgrade.UpgradeManagedFields(&initialObject, "kubectl-client-side-apply", "kubectl", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectedYAML := []byte(`
+apiVersion: v1
+data:
+  key: value
+  legacy: unused
+kind: ConfigMap
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
```

Comment:
Should we also clean-up the annotation to make it clear that this isn't reversible?

Reply:
Does this annotation only apply to kubectl? I guess it should only be wiped if `csaManagerName` is `kubectl-client-side-apply`?

Reply:
No I think it should be wiped no matter what. This is strictly a left-over from client-side apply. And if you're running the migration, you should NOT use client-side apply, and this annotation calls for it.